### PR TITLE
[REEF-1389] Add wrapper interface for DirectoryInfo and FileInfo to f…

### PR DESCRIPTION
**Pull request by afchung**
…acilitate testing for REEF-1357

JIRA:
  [REEF-1389](https://issues.apache.org/jira/browse/REEF-1389)

### CODE REVIEW

**Comment by dkm2110** [comment]
@afchung I am wondering what are the relation of these interfaces to the IFileSystem interface we already have. In some sense shouldn't IDirectoryInfo and IFileInfo be part of general IFileSystem interface and we just need to expand it...Or I am over generalizing? Basically, if a user comes to REEF, and looks at these interfaces will he be clear what IFileSystem is intended for and what IFileInfo and other related interfaces intended for. I will do the review once we figure out this question. @markusweimer can you also comment?

**Comment by afchung** [comment]
@dkm2110 these interfaces are simply to allow us to perform tests by mocking out `DirectoryInfo` and `FileInfo`. These, at the moment, very specifically refer to local directories and files only, and many of these concepts/methods may not apply to remote file systems. We can in time certainly extend them for usage in `IFileSystem`, but supporting such interfaces will not be trivial for remote file systems.

File: lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+        }
+
+        public DateTime LastAccessTime
+        {
+            get { return _directoryInfo.LastAccessTime; }
+        }
+
+        public DateTime LastAccessTimeUtc
+        {
+            get { return _directoryInfo.LastAccessTime; }
```
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs]: _directoryInfo.LastAccessTimeUtc?

File: lang/cs/Org.Apache.REEF.IO/Files/IFIleSystemInfo.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IO/Files/IFIleSystemInfo.cs]: IFIleSystemInfo.cs -> IFileSystemInfo.cs

File: lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        {
+            return ConvertDirectoryArray(_directoryInfo.GetDirectories(searchPattern));
+        }
+
+        public IDirectoryInfo[] GetDirectories(string searchPattern, SearchOption searchOption)
+        {
+            return ConvertDirectoryArray(_directoryInfo.GetDirectories(searchPattern, searchOption));
+        }
+
+        private static IDirectoryInfo[] ConvertDirectoryArray(DirectoryInfo[] directoryInfos)
```
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs]: Should declare Private function after all public ones.

File: lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+        {
+            return ConvertFileArray(_directoryInfo.GetFiles(searchPattern));
+        }
+
+        public IFileInfo[] GetFiles(string searchPattern, SearchOption searchOption)
+        {
+            return ConvertFileArray(_directoryInfo.GetFiles(searchPattern, searchOption));
+        }
+
+        private static IFileInfo[] ConvertFileArray(FileInfo[] fileInfos)
```
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs]: Should be after all public functions.

File: lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        {
+            return ConvertFileSystemInfoArray(_directoryInfo.GetFileSystemInfos(searchPattern));
+        }
+
+        public IFileSystemInfo[] GetFileSystemInfos(string searchPattern, SearchOption searchOption)
+        {
+            return ConvertFileSystemInfoArray(_directoryInfo.GetFileSystemInfos(searchPattern, searchOption));
+        }
+
+        private static IFileSystemInfo[] ConvertFileSystemInfoArray(IEnumerable<FileSystemInfo> fsInfos)
```
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs]: Should be after all public functions.

**Comment by dkm2110** [comment]
@afchung Left some minor comments. Apart from that two more points:
a) There is no documentation currently for any public functions in interface and implementations.
b) If we think that these interfaces and implementations will be used only in tests in future we can move it to test project also.

**Comment by afchung** [comment]
@dkm2110 for b), these interfaces and implementations will not only be used in tests, but will be used to facilitate testing. For that to work, we need to integrate it into our public facing APIs.

For example when the user wants to use an `IFileInfo` object in our APIs, they will simply do:
`IFileInfo info = DefaultFileInfo.FromFileInfo(new FileInfo(...));`

For our testing, because we don't actually want to use a `FileInfo` object, we can use NSubstitute in the following way:
```
IFileInfo infoForTesting = Substitute.For<IFileInfo>();
infoForTesting.DirectoryName.Returns("FakeDirectoryName");
```

Such that when we test one of our functions with the parameter of an `IFileInfo` with the fake `infoForTesting` object, it will not perform actual actions on a file. This requires our functions to expose an interface instead of a class, since NSubstitute does not work on concrete classes without virtual methods such as `FileInfo`<sup>[0](http://nsubstitute.github.io/help/creating-a-substitute/)</sup><sup>[1](http://stackoverflow.com/questions/17182355/can-i-get-nsubstitute-to-auto-mock-my-concrete-classes)</sup>.

**Comment by afchung** [comment]
@dkm2110 I've addressed your comments, please have another look. Thanks!

**Comment by dkm2110** [comment]
@afchung basically I learnt something new i.e. Nsubstitute :) ... I will test and merge later tonight.

**Comment by afchung** [comment]
@dkm2110 what is the status on this?

**Comment by dkm2110** [comment]
Testing now.

File: lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+        private DefaultDirectoryInfo(DirectoryInfo info)
+        {
+            _directoryInfo = info;
+        }
+
+        /// <summary>
+        /// Factory method to create an <see cref="IDirectoryInfo"/> object
+        /// from a <see cref="DirectoryInfo"/> object.
+        /// </summary>
+        public static IDirectoryInfo FromDirectoryInfo(DirectoryInfo info)
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs]: Where is `DirectoryInfo` defined? is this a standard C# class?

File: lang/cs/Org.Apache.REEF.IO/Files/DefaultFileInfo.cs [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+        private DefaultFileInfo(FileInfo info)
+        {
+            _fileInfo = info;
+        }
+
+        /// <summary>
+        /// Factory method to create an <see cref="IFileInfo"/> object
+        /// from a <see cref="FileInfo"/> object.
+        /// </summary>
+        public static IFileInfo FromFileInfo(FileInfo fileInfo)
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/Files/DefaultFileInfo.cs]: Same as above: Where is `FileInfo` defined?

**Comment by markusweimer** [comment]
I'm also confused about the role of these new classes. Do they add interfaces on top of APIs we already have? If so, all uses of the classes should be replaced with uses of the interfaces.

File: lang/cs/Org.Apache.REEF.IO/Files/DefaultFileInfo.cs (hunk 7 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.IO/Files/DefaultFileInfo.cs]: Yes, this is a standard C# class.

File: lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs (hunk 6 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.IO/Files/DefaultDirectoryInfo.cs]: Yes, this is a standard C# class.

**Comment by afchung** [comment]
@markusweimer They do not add interfaces on top of APIs we already have. They are a precursor to allow #994 to be unit-testable. For more explanation, see my reply do @dkm2110 above. Thanks!

**Comment by markusweimer** [comment]
Is the following a fair description of this change?

> This adds interface for classes in the .NET standard library such that we can mock implementations of them for our unit tests.

**Comment by afchung** [comment]
@markusweimer yes, should I reflect that message somewhere?

**Comment by markusweimer** [comment]
I've copied this description to the JIRA. In case we ever wonder again :)

**Pull request closed without merging**